<!DOCTYPE html>


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
  
    <title>HiC-Pro: An optimized and flexible pipeline for Hi-C processing. &mdash; HiC-Pro 2.3.1 documentation</title>
  <!-- htmltitle is before nature.css - we use this hack to load bootstrap first -->
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <link rel="stylesheet"
        href="_static/css/bootstrap.min.css"
	media="screen" />
    <script src="_static/js/jquery.js"
          type="text/javascript"></script>
  <script src="_static/js/bootstrap.min.js"
          type="text/javascript"></script>


    
    <link rel="stylesheet" href="_static/basic.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    './',
        VERSION:     '2.3.1',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <link rel="top" title="HiC-Pro 2.3.1 documentation" href="index.html" />
    <link rel="next" title="FAQ" href="FAQ.html" />
    <link rel="prev" title="User cases" href="USER_CASES.html" />
<meta charset='utf-8'>
<meta http-equiv='X-UA-Compatible' content='IE=edge,chrome=1'>
<meta name='viewport'
      content='width=device-width, initial-scale=1.0, maximum-scale=1'>
<meta name="apple-mobile-web-app-capable" content="yes">
<link rel="stylesheet"
        href="_static/css/jhepc.css"
         media="screen" />
  <link href='http://fonts.googleapis.com/css?family=PT+Serif'
        rel='stylesheet' type='text/css'>
  <link href='http://fonts.googleapis.com/css?family=PT+Sans'
        rel='stylesheet' type='text/css'>
  <link href='http://fonts.googleapis.com/css?family=PT+Sans+Narrow'
        rel='stylesheet' type='text/css'>



  </head>
  <body><div class="navbar">
    <a class="navbar-brand" href="index.html">
      
    </a>
    <ul class="nav navbar-nav">

    </ul>
  </div>
    <div class="container">
  
  <div class="section" id="compatibility-with-other-software">
<span id="compatibility"></span><h1>Compatibility with other software<a class="headerlink" href="#compatibility-with-other-software" title="Permalink to this headline">¶</a></h1>
<div class="line-block">
<div class="line">HiC-Pro is designed to process Hi-C data from .fastq files to normalized contact matrices. Additional software or methods should then be used to perform downstream analysis. We therefore decided to propose several utilities to increase the compatibility of HiC-Pro with other tools. Most of the scripts presented below were developed by HiC-Pro users or collaborators. Many thanks to them for sharing their work with us, and do not hesitate the contact us with many new ideas !</div>
</div>
<div class="section" id="juicebox-visualization-software">
<h2>Juicebox visualization software<a class="headerlink" href="#juicebox-visualization-software" title="Permalink to this headline">¶</a></h2>
<p>The <a href="#id2"><span class="problematic" id="id3">`Juicebox&lt;http://aidenlab.org/juicebox/&gt;`_</span></a> software allows dynamic visualization of HiC contact maps. The list of valid interaction products called by HiC-Pro can can be converted in Juicebox input format (.hic file). To do so, please have a look at the hicpro2juicebox.sh utility.</p>
</div>
<div class="section" id="hicplotter-visualition-software">
<h2>HiCPlotter visualition software<a class="headerlink" href="#hicplotter-visualition-software" title="Permalink to this headline">¶</a></h2>
<p>The output format of HiC-Pro can be loaded into the <a href="#id4"><span class="problematic" id="id5">`HiCPlotter&lt;https://github.com/kcakdemir/HiCPlotter&gt;`_</span></a> software for Hi-C data visualization.
Here is a small example of how to use it.</p>
<div class="highlight-guess"><div class="highlight"><pre><span class="cp">## Plot the genome-wide map at 1Mb resolution</span>
<span class="n">python</span> <span class="n">HiCPlotter</span><span class="p">.</span><span class="n">py</span> <span class="o">-</span><span class="n">f</span> <span class="n">hic_results</span><span class="o">/</span><span class="n">matrix</span><span class="o">/</span><span class="n">sample1</span><span class="o">/</span><span class="n">iced</span><span class="o">/</span><span class="mi">1000000</span><span class="o">/</span><span class="n">sample1_1000000_iced</span><span class="p">.</span><span class="n">matrix</span> <span class="o">-</span><span class="n">o</span> <span class="n">Examplegw</span> <span class="o">-</span><span class="n">r</span> <span class="mi">1000000</span> <span class="o">-</span><span class="n">tri</span> <span class="mi">1</span> <span class="o">-</span><span class="n">bed</span> <span class="n">hic_results</span><span class="o">/</span><span class="n">matrix</span><span class="o">/</span><span class="n">sample1</span><span class="o">/</span><span class="n">raw</span><span class="o">/</span><span class="mi">1000000</span><span class="o">/</span><span class="n">sample1_1000000_ord</span><span class="p">.</span><span class="n">bed</span> <span class="o">-</span><span class="n">n</span> <span class="n">hES</span> <span class="o">-</span><span class="n">wg</span> <span class="mi">1</span> <span class="o">-</span><span class="n">chr</span> <span class="n">chrX</span>

<span class="cp">## Plot the chrX at 150Kb resolution</span>
<span class="n">python</span> <span class="n">HiCPlotter</span><span class="p">.</span><span class="n">py</span> <span class="o">-</span><span class="n">f</span> <span class="n">hic_results</span><span class="o">/</span><span class="n">matrix</span><span class="o">/</span><span class="n">sample1</span><span class="o">/</span><span class="n">iced</span><span class="o">/</span><span class="mi">150000</span><span class="o">/</span><span class="n">sample1_150000_iced</span><span class="p">.</span><span class="n">matrix</span> <span class="o">-</span><span class="n">o</span> <span class="n">Exemple</span> <span class="o">-</span><span class="n">r</span> <span class="mi">150000</span> <span class="o">-</span><span class="n">tri</span> <span class="mi">1</span> <span class="o">-</span><span class="n">bed</span> <span class="n">hic_results</span><span class="o">/</span><span class="n">matrix</span><span class="o">/</span><span class="n">sample1</span><span class="o">/</span><span class="n">raw</span><span class="o">/</span><span class="mi">150000</span><span class="o">/</span><span class="n">sample1_150000_ord</span><span class="p">.</span><span class="n">bed</span> <span class="o">-</span><span class="n">n</span> <span class="n">Test</span> <span class="o">-</span><span class="n">chr</span> <span class="n">chrX</span> <span class="o">-</span><span class="n">ptr</span> <span class="mi">1</span>
</pre></div>
</div>
</div>
<div class="section" id="tads-calling-directionality-index">
<h2>TADs calling (Directionality Index)<a class="headerlink" href="#tads-calling-directionality-index" title="Permalink to this headline">¶</a></h2>
<p>TADs calling is frequently run on Hi-C data at a 40Kb or higher resolution. In order to run the first method proposed by Dixon et al., you can generated the input file using the sparseTodense.py utility.</p>
</div>
<div class="section" id="fit-hi-c">
<h2>Fit-Hi-C<a class="headerlink" href="#fit-hi-c" title="Permalink to this headline">¶</a></h2>
<p><a class="reference external" href="https://bioconductor.org/packages/release/bioc/html/FitHiC.html">Fit-Hi-C</a> is a popular approach to detect significant interaction on intra-chromosoml contact map.
Please use the hicpro2fithic.py utility to connvert HiC-Pro output to Fit-Hi-C input files.</p>
</div>
<div class="section" id="r-and-hitc-bioconductor-package">
<h2>R and HiTC BioConductor package<a class="headerlink" href="#r-and-hitc-bioconductor-package" title="Permalink to this headline">¶</a></h2>
<p>The <a class="reference external" href="https://bioconductor.org/packages/release/bioc/html/HiTC.html">HiTC</a> package proposes function to import HiC-Pro output into the R environment.
Contact matrices are then stored as sparse Matrix, which can be used for any downstream analysis.</p>
</div>
</div>


    </div>
<footer class="footer">
    <p class="pull-right">
      <a href="#">Back to top</a>
      
    </p>
    <p>
        &copy; Copyright 2015, Nicolas Servant.<br/>
      Created using <a href="http://sphinx.pocoo.org/">Sphinx</a> 1.2.2.<br/>
    </p>
</footer>
  </body>
</html>