<!DOCTYPE html>


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
  
    <title>HiC-Pro: An optimized and flexible pipeline for Hi-C processing. &mdash; HiC-Pro 2.3.1 documentation</title>
  <!-- htmltitle is before nature.css - we use this hack to load bootstrap first -->
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <link rel="stylesheet"
        href="_static/css/bootstrap.min.css"
	media="screen" />
    <script src="_static/js/jquery.js"
          type="text/javascript"></script>
  <script src="_static/js/bootstrap.min.js"
          type="text/javascript"></script>


    
    <link rel="stylesheet" href="_static/basic.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    './',
        VERSION:     '2.3.1',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <link rel="top" title="HiC-Pro 2.3.1 documentation" href="index.html" />
    <link rel="next" title="Allele specific analysis" href="AS.html" />
    <link rel="prev" title="HiC-Pro Quick Start Guide" href="QUICKSTART.html" />
<meta charset='utf-8'>
<meta http-equiv='X-UA-Compatible' content='IE=edge,chrome=1'>
<meta name='viewport'
      content='width=device-width, initial-scale=1.0, maximum-scale=1'>
<meta name="apple-mobile-web-app-capable" content="yes">
<link rel="stylesheet"
        href="_static/css/jhepc.css"
         media="screen" />
  <link href='http://fonts.googleapis.com/css?family=PT+Serif'
        rel='stylesheet' type='text/css'>
  <link href='http://fonts.googleapis.com/css?family=PT+Sans'
        rel='stylesheet' type='text/css'>
  <link href='http://fonts.googleapis.com/css?family=PT+Sans+Narrow'
        rel='stylesheet' type='text/css'>



  </head>
  <body role="document"><div class="navbar">
    <a class="navbar-brand" href="index.html">
      
    </a>
    <ul class="nav navbar-nav">

    </ul>
  </div>
    <div class="container">
  
  <div class="section" id="hic-pro-manual">
<h1>HiC-Pro Manual<a class="headerlink" href="#hic-pro-manual" title="Permalink to this headline">¶</a></h1>
<p>Modified - 09th September 2015
Reference version - HiC-Pro 2.6.1</p>
<div class="section" id="setting-the-configuration-file">
<h2>Setting the configuration file<a class="headerlink" href="#setting-the-configuration-file" title="Permalink to this headline">¶</a></h2>
<ol class="arabic simple">
<li>Copy and edit the configuration file <em>&#8216;config-hicpro.txt&#8217;</em> in your local folder. The &#8216;[]&#8217; options are optional and can be undefined.</li>
</ol>
<table border="1" class="docutils">
<colgroup>
<col width="29%" />
<col width="71%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head" colspan="2">SET UP SYSTEM AND PBS/TORQUE MODE</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>N_CPU</td>
<td>Number of CPU allows per job</td>
</tr>
<tr class="row-odd"><td>LOGFILE</td>
<td>Name of the main log file</td>
</tr>
<tr class="row-even"><td>[PBS_SUFFIX]</td>
<td>Name of PBS/Torque job on the custer</td>
</tr>
<tr class="row-odd"><td>[PBS_MEM]</td>
<td>Memory (RAM) required per job</td>
</tr>
<tr class="row-even"><td>[PBS_WALLTIME]</td>
<td>WallTime allows per job</td>
</tr>
<tr class="row-odd"><td>[PBS_MAIL]</td>
<td>User mail for PBS/Torque report</td>
</tr>
</tbody>
</table>
<hr class="docutils" />
<table border="1" class="docutils">
<colgroup>
<col width="17%" />
<col width="83%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head" colspan="2">DATA AND READS ALIGNMENT OPTIONS</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>PAIR1_EXT</td>
<td>Keyword for first mate detection. <em>Default:_R1</em></td>
</tr>
<tr class="row-odd"><td>PAIR2_EXT</td>
<td>Keywoard for seconde mate detection. <em>Default:_R2</em></td>
</tr>
<tr class="row-even"><td>FORMAT</td>
<td>Sequencing qualities encoding. <em>Default: phred33</em></td>
</tr>
<tr class="row-odd"><td>MIN_MAPQ</td>
<td>Minimum mapping quality. Reads with lower quality are discarded. <em>Default: 0</em></td>
</tr>
<tr class="row-even"><td>BOWTIE2_IDX_PATH</td>
<td>Path to bowtie2 indexes</td>
</tr>
<tr class="row-odd"><td>BOWTIE2_GLOBAL_OPTIONS</td>
<td>bowtie2 options for mapping step1. <em>Default: &#8211;very-sensitive -L 30 &#8211;score-min L,-0.6,-0.2 &#8211;end-to-end &#8211;reorder</em></td>
</tr>
<tr class="row-even"><td>BOWTIE2_LOCAL_OPTIONS</td>
<td>bowtie2 options for mapping step2. <em>Default: &#8211;very-sensitive -L 20 &#8211;score-min L,-0.6,-0.2 &#8211;end-to-end &#8211;reorder</em></td>
</tr>
</tbody>
</table>
<hr class="docutils" />
<table border="1" class="docutils">
<colgroup>
<col width="14%" />
<col width="86%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head" colspan="2">ANNOTATION FILES</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>REFERENCE_GENOME</td>
<td>Reference genome prefix used for genome indexes. <em>Default: hg19</em></td>
</tr>
<tr class="row-odd"><td>GENOME_SIZE</td>
<td>Chromsome size file. Loaded from the ANNOTATION folder in the HiC-Pro installation directory. <em>Default: chrom_hg19.sizes</em></td>
</tr>
</tbody>
</table>
<hr class="docutils" />
<table border="1" class="docutils">
<colgroup>
<col width="13%" />
<col width="2%" />
<col width="85%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head" colspan="3">ALLLELE SPECIFIC ANALYSIS</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td colspan="3">[ALLELE_SPECIFIC_SNP]     | VCF file to SNPs which can be used to distinguish parental origin. See the <a class="reference external" href="AS.rst">allele specific section</a> for more details</td>
</tr>
</tbody>
</table>
<hr class="docutils" />
<table border="1" class="docutils">
<colgroup>
<col width="14%" />
<col width="86%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head" colspan="2">DIGESTION Hi-C</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>[GENOME_FRAGMENT]</td>
<td>BED file with restriction fragments. Full path or name of file available in the ANNOTATION folder. <em>Example: HindIII_resfrag_hg19.bed</em></td>
</tr>
<tr class="row-odd"><td>[LIGATION SITE]</td>
<td>Ligation site sequence used for reads trimming. Depends on the fill in strategy. <em>Example: AAGCTAGCTT</em></td>
</tr>
<tr class="row-even"><td>[MIN_FRAG_SIZE]</td>
<td>Maximum size of restriction fragments to consider for the Hi-C processing. <em>Example: 100</em></td>
</tr>
<tr class="row-odd"><td>[MAX_FRAG_SIZE]</td>
<td>Maximum size of restriction fragments to consider for the Hi-C processing. <em>Example: 100000</em></td>
</tr>
<tr class="row-even"><td>[MIN_INSERT_SIZE]</td>
<td>Minimum sequenced insert size. Shorter 3C products are discarded. <em>Example: 100</em></td>
</tr>
<tr class="row-odd"><td>[MAX_INSERT_SIZE]</td>
<td>Maximum sequenced insert size. Larger 3C products are discarded. <em>Example: 600</em></td>
</tr>
</tbody>
</table>
<hr class="docutils" />
<table border="1" class="docutils">
<colgroup>
<col width="19%" />
<col width="81%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head" colspan="2">Hi-C PROCESSING</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>[MIN_CIS_DIST]</td>
<td>Filter short range contact below the specified distance. Mainly useful for DNase Hi-C. <em>Example: 1000</em></td>
</tr>
<tr class="row-odd"><td>GET_ALL_INTERACTION_CLASSES</td>
<td>Create output files with all classes of 3C products. <em>Default: 0</em></td>
</tr>
<tr class="row-even"><td>GET_PROCESS_BAM</td>
<td>Create a BAM file with all aligned reads flagged according to their classifaction and mapping category. <em>Default: 0</em></td>
</tr>
<tr class="row-odd"><td>RM_SINGLETON</td>
<td>Remove singleton reads. <em>Default: 1</em></td>
</tr>
<tr class="row-even"><td>RM_MULTI</td>
<td>Remove multi-mapped reads. <em>Default: 1</em></td>
</tr>
<tr class="row-odd"><td>RM_DUP</td>
<td>Remove duplicated reads&#8217; pairs. <em>Default: 1</em></td>
</tr>
</tbody>
</table>
<hr class="docutils" />
<table border="1" class="docutils">
<colgroup>
<col width="19%" />
<col width="81%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head" colspan="2">GENOME-WIDE CONTACT MAPS</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>BIN_SIZE</td>
<td>Resolution of contact maps to generate (space separated). <em>Default: 20000 40000 150000 500000 1000000</em></td>
</tr>
<tr class="row-odd"><td>BIN_STEP</td>
<td>Binning step size in ‘n’ coverage _i.e._ window step. <em>Default: 1</em></td>
</tr>
<tr class="row-even"><td>MATRIX_FORMAT</td>
<td>Output matrix format. Must be complete, asis, upper or lower. <em>Default: upper</em></td>
</tr>
</tbody>
</table>
<hr class="docutils" />
<table border="1" class="docutils">
<colgroup>
<col width="19%" />
<col width="81%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head" colspan="2">NORMALIZATION</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>MAX_ITER</td>
<td>Maximum number of iteration for ICE normalization. <em>Default: 100</em></td>
</tr>
<tr class="row-odd"><td>SPARSE_FILTERING</td>
<td>Define which pourcentage of bins with high sparsity should be force to zero. <em>Default: 0.02</em></td>
</tr>
<tr class="row-even"><td>EPS</td>
<td>The relative increment in the results before declaring convergence. <em>Default: 0.1</em></td>
</tr>
</tbody>
</table>
</div>
<hr class="docutils" />
<div class="section" id="run-hic-pro-in-sequential-mode">
<h2>Run HiC-Pro in sequential mode<a class="headerlink" href="#run-hic-pro-in-sequential-mode" title="Permalink to this headline">¶</a></h2>
<p>HiC-Pro can be run in a step-by-step mode.
Available steps are described in the help command.</p>
<div class="highlight-guess"><div class="highlight"><pre>HiC-Pro --help
usage : HiC-Pro -i INPUT -o OUTPUT -c CONFIG [-s ANALYSIS_STEP] [-p] [-h] [-v]
Use option -h|--help for more information

HiC-Pro 2.6.0
---------------
OPTIONS

 -i|--input INPUT : input data folder; Must contains a folder per sample with input files
 -o|--output OUTPUT : output folder
 -c|--conf CONFIG : configuration file for Hi-C processing
 [-p|--parallel] : if specified run HiC-Pro in PBS/Torque mode
 [-s|--step ANALYSIS_STEP] : run only a subset of the HiC-Pro workflow; if not specified the complete workflow is run
    mapping: perform reads alignment
    proc_hic: perform Hi-C filtering
    quality_checks: run Hi-C quality control plots
    build_contact_maps: build raw inter/intrachromosomal contact maps
    ice_norm: run ICE normalization on contact maps
 [-h|--help]: help
 [-v|--version]: version
</pre></div>
</div>
<p>As an exemple, if you want to only want to only align the sequencing reads and run a quality control, use :</p>
<div class="highlight-guess"><div class="highlight"><pre>MY_INSTALL_PATH/bin/HiC-Pro -i FULL_PATH_TO_RAW_DATA -o FULL_PATH_TO_OUTPUTS -c MY_LOCAL_CONFIG_FILE -s mapping -s quality_checks
</pre></div>
</div>
<p>Note that in sequential mode, the INPUT argument depends on the analysis step. See te <a class="reference external" href="USER_CASES.rst">user&#8217;s cases</a> for more examples.</p>
<table border="1" class="docutils">
<colgroup>
<col width="53%" />
<col width="47%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head" colspan="2">INPUT DATA TYPE IN STEPWISE MODE</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>-s mapping</td>
<td>.fastq(.gz) files</td>
</tr>
<tr class="row-odd"><td>-s proc_hic</td>
<td>.bam files</td>
</tr>
<tr class="row-even"><td>-s quality_checks</td>
<td>.bam files</td>
</tr>
<tr class="row-odd"><td>-s build_contact_maps</td>
<td>.validPairs files</td>
</tr>
<tr class="row-even"><td>-s ice_norm</td>
<td>.matrix files</td>
</tr>
</tbody>
</table>
</div>
<div class="section" id="how-does-hic-pro-work">
<h2>How does HiC-Pro work ?<a class="headerlink" href="#how-does-hic-pro-work" title="Permalink to this headline">¶</a></h2>
<p>The HiC-Pro workflow can be divided in five main steps presented below.</p>
<div class="figure">
<a class="reference internal image-reference" href="_images/hicpro_wkflow.png"><img alt="_images/hicpro_wkflow.png" src="_images/hicpro_wkflow.png" /></a>
</div>
<ol class="arabic simple">
<li><strong>Reads Mapping</strong></li>
</ol>
<div class="line-block">
<div class="line">Each mate is independantly aligned on the reference genome. The mapping is performed in two steps. First, the reads are aligned using an end-to-end aligner. Second, reads spanning the ligation junction are trimmmed from their 3&#8217; end, and aligned back on the genome. Aligned reads for both fragment mates are then paired in a single paired-end BAM file. Singletons and multi-hits can be discarded according the confirguration parameters. Note that if if the <em>LIGATION_SITE</em> parameter in the not defined, HiC-Pro will skip the second step of mapping.</div>
</div>
<ol class="arabic simple" start="2">
<li><strong>Fragment assignment and filtering</strong></li>
</ol>
<div class="line-block">
<div class="line">Each aligned reads can be assigned to one restriction fragment according to the reference genome and the restriction enzyme.</div>
<div class="line">The next step is to separate the invalid ligation products from the valid pairs. Dangling end and self circles pairs are therefore excluded.</div>
<div class="line">Only valid pairs involving two different restriction fragments are used to build the contact maps. Duplicated valid pairs associated to PCR artefacts are discarded.</div>
<div class="line">The fragment assignment can be visualized through a BAM files of aliged pairs where each pair is flagged according to its classification.</div>
<div class="line">In case of Hi-C protocols that do not require a restriction enzyme such as DNase Hi-C or micro Hi-C, the assignment to a restriction is not possible. If no <em>GENOME_FRAGMENT</em> file are specified, this step is ignored. Short range interactions can however still be discarded using the <em>MIN_CIS_DIST</em> parameter.</div>
</div>
<ol class="arabic simple" start="3">
<li><strong>Quality Controls</strong></li>
</ol>
<div class="line-block">
<div class="line">HiC-Pro performs a couple of quality controls for most of the analysis steps. The alignment statistics are the first quality controls. Aligned reads in the first (end-to-end) step, and alignment after trimming are reported. Note that in pratice, we ususally observed around 10-20% of trimmed reads. An abnormal level of trimmed reads can reflect a ligation issue.</div>
<div class="line">Once the reads are aligned on the genome, HiC-pro checks the number of singleton, multiple hits or duplicates. The fraction of valid pairs are presented for each type of ligation products. Invalid pairs such as dangling and or self-circle are also represented. A high level of dangling ends, or an imbalance in valid pairs ligation type can be due to a ligation, fill-in or digestion issue.</div>
<div class="line">Finally HiC-Pro also calculated the distribution of fragment size on a subset of valid pairs. Additional statistics will report the fraction of intra/inter-chromosomal contacts, as well as the proportion of short range (&lt;20kb) versus long range (&gt;20kb) contacts.</div>
</div>
<ol class="arabic simple" start="4">
<li><strong>Map builder</strong></li>
</ol>
<div class="line-block">
<div class="line">Intra et inter-chromosomal contact maps are build for all specified resolutions. The genome is splitted into bins of equal size. Each valid interaction is associated with the genomic bins to generate the raw maps.</div>
</div>
<ol class="arabic simple" start="5">
<li><strong>ICE normalization</strong></li>
</ol>
<div class="line-block">
<div class="line">Hi-C data can contain several sources of biases which has to be corrected. HiC-Pro proposes a fast implementation of the original ICE normalization algorithm (Imakaev et al. 2012), making the assumption of equal visibility of each fragment. The ICE normalization can be used as a standalone python package through the <a class="reference external" href="https://github.com/hiclib/">iced python package</a>.</div>
</div>
</div>
<div class="section" id="browsing-the-results">
<h2>Browsing the results<a class="headerlink" href="#browsing-the-results" title="Permalink to this headline">¶</a></h2>
<p>All outputs follow the input organization, with one folder per sample.
See the <a class="reference external" href="RESULTS.rst">Results section</a> for more information.</p>
<ul class="simple">
<li><em>bowtie_results</em></li>
</ul>
<p>The <em>bowtie_results</em> folder contains the results of the reads mapping. The results of first mapping step are available in the <em>bwt2_glob</em> folder, and the seconnd step in the <em>bwt2_loc</em> folder. Final BAM files, reads pairing, and mapping statistics are available on the <em>bwt2</em> folder.</p>
<ul class="simple">
<li><em>hic_results</em></li>
</ul>
<div class="line-block">
<div class="line">This folder contains all Hi-C processed data, and is further divided in several sub-folders.</div>
<div class="line">The <em>data</em> folder is used to store the valid interaction products (<em>.validPairs</em>), as well as other statisics files.</div>
<div class="line">The contact maps are then available in the <em>matrix</em> folder. The <em>matrix</em> folder is organized with <em>raw</em> and <em>iced</em> contact maps for all resolutions.</div>
<div class="line">Finally, the <em>pic</em> folder contains graphical outputs of the quality control checks.</div>
</div>
</div>
</div>


    </div>
<footer class="footer">
    <p class="pull-right">
      <a href="#">Back to top</a>
      
    </p>
    <p>
        &copy; Copyright 2015, Nicolas Servant.<br/>
      Created using <a href="http://sphinx.pocoo.org/">Sphinx</a> 1.3.1.<br/>
    </p>
</footer>
  </body>
</html>