<!DOCTYPE html>


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
  
    <title>HiC-Pro: An optimized and flexible pipeline for Hi-C processing. &mdash; HiC-Pro 2.3.1 documentation</title>
  <!-- htmltitle is before nature.css - we use this hack to load bootstrap first -->
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <link rel="stylesheet"
        href="_static/css/bootstrap.min.css"
	media="screen" />
    <script src="_static/js/jquery.js"
          type="text/javascript"></script>
  <script src="_static/js/bootstrap.min.js"
          type="text/javascript"></script>


    
    <link rel="stylesheet" href="_static/basic.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    './',
        VERSION:     '2.3.1',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <link rel="top" title="HiC-Pro 2.3.1 documentation" href="index.html" />
    <link rel="next" title="ALLELE-SPECIFIC ANALYSIS" href="AS.html" />
    <link rel="prev" title="HiC-Pro Quick Start Guide" href="QUICKSTART.html" />
<meta charset='utf-8'>
<meta http-equiv='X-UA-Compatible' content='IE=edge,chrome=1'>
<meta name='viewport'
      content='width=device-width, initial-scale=1.0, maximum-scale=1'>
<meta name="apple-mobile-web-app-capable" content="yes">
<link rel="stylesheet"
        href="_static/css/jhepc.css"
         media="screen" />
  <link href='http://fonts.googleapis.com/css?family=PT+Serif'
        rel='stylesheet' type='text/css'>
  <link href='http://fonts.googleapis.com/css?family=PT+Sans'
        rel='stylesheet' type='text/css'>
  <link href='http://fonts.googleapis.com/css?family=PT+Sans+Narrow'
        rel='stylesheet' type='text/css'>



  </head>
  <body><div class="navbar">
    <a class="navbar-brand" href="index.html">
      
    </a>
    <ul class="nav navbar-nav">

    </ul>
  </div>
    <div class="container">
  
  <div class="section" id="hic-pro-manual">
<h1>HiC-Pro Manual<a class="headerlink" href="#hic-pro-manual" title="Permalink to this headline">¶</a></h1>
<div class="section" id="setting-the-configuration-file">
<h2>Setting the configuration file<a class="headerlink" href="#setting-the-configuration-file" title="Permalink to this headline">¶</a></h2>
<ol class="arabic simple">
<li>Copy and edit the configuration file <em>&#8216;config-hicpro.txt&#8217;</em> in your local folder. The &#8216;[&#8216; options are optional and can be undefined.</li>
</ol>
<table border="1" class="docutils">
<colgroup>
<col width="27%" />
<col width="0%" />
<col width="73%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head" colspan="3">SET UP SYSTEM AND PBS/TORQUE MODE</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td colspan="2">N_CPU</td>
<td>Number of CPU allows fper job</td>
</tr>
<tr class="row-odd"><td colspan="2">LOGFILE</td>
<td>Name of the main log file</td>
</tr>
<tr class="row-even"><td colspan="2">[PBS_SUFFIX]</td>
<td>Name of PBS/Torque job on the custer</td>
</tr>
<tr class="row-odd"><td colspan="2">[PBS_MEM]</td>
<td>Memory (RAM) required per job</td>
</tr>
<tr class="row-even"><td colspan="2">[PBS_WALLTIME]</td>
<td>WallTime allows per job</td>
</tr>
<tr class="row-odd"><td colspan="2">[PBS_MAIL]</td>
<td>User mail for PBS/Torque report</td>
</tr>
</tbody>
</table>
<hr class="docutils" />
<table border="1" class="docutils">
<colgroup>
<col width="17%" />
<col width="83%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head" colspan="2">READS ALIGNMENT OPTIONS</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>RAW_DIR</td>
<td>Link to rawdata folder. The user usually not need to change this option. <em>Default: rawdata</em></td>
</tr>
<tr class="row-odd"><td>PAIR1_EXT</td>
<td>Keyword for first mate detection. <em>Default:_R1</em></td>
</tr>
<tr class="row-even"><td>PAIR2_EXT</td>
<td>Keywoard for seconde mate detection. <em>Default:_R2</em></td>
</tr>
<tr class="row-odd"><td>FORMAT</td>
<td>Sequencing qualities encoding. <em>Default: phred33</em></td>
</tr>
<tr class="row-even"><td>MIN_MAPQ</td>
<td>Minimum mapping quality. Reads with lower quality are discarded. <em>Default: 0</em></td>
</tr>
<tr class="row-odd"><td>LIGATION SITE</td>
<td>Ligation site sequence used for reads trimming. Depends on the fill in strategy. <em>Default: AAGCTAGCTT</em></td>
</tr>
<tr class="row-even"><td>BOWTIE2_IDX_PATH</td>
<td>Path to bowtie2 indexes</td>
</tr>
<tr class="row-odd"><td>BOWTIE2_GLOBAL_OPTIONS</td>
<td>bowtie2 options for mapping step1. <em>Default: &#8211;very-sensitive -L 30 &#8211;score-min L,-0.6,-0.2 &#8211;end-to-end &#8211;reorder</em></td>
</tr>
<tr class="row-even"><td>BOWTIE2_LOCAL_OPTIONS</td>
<td>bowtie2 options for mapping step2. <em>Default: &#8211;very-sensitive -L 20 &#8211;score-min L,-0.6,-0.2 &#8211;end-to-end &#8211;reorder</em></td>
</tr>
</tbody>
</table>
<hr class="docutils" />
<hr class="docutils" />
<table border="1" class="docutils">
<colgroup>
<col width="19%" />
<col width="81%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head" colspan="2">Hi-C PROCESSING</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>[MIN_INSERT_SIZE]</td>
<td>Minimum sequenced insert size. Shorter 3C products are discarded</td>
</tr>
<tr class="row-odd"><td>[MAX_INSERT_SIZE]</td>
<td>Maximum sequenced insert size. Larger 3C products are discarded</td>
</tr>
<tr class="row-even"><td>GET_ALL_INTERACTION_CLASSES</td>
<td>Create output files with all classes of 3C products. <em>Default: 1</em></td>
</tr>
<tr class="row-odd"><td>GET_PROCESS_BAM</td>
<td>Create a BAM file with all aligned reads flagged according to their classifaction and mapping category. <em>Default: 0</em></td>
</tr>
<tr class="row-even"><td>RM_SINGLETON</td>
<td>Remove singleton reads. <em>Default: 1</em></td>
</tr>
<tr class="row-odd"><td>RM_MULTI</td>
<td>Remove multi-mapped reads. <em>Default: 1</em></td>
</tr>
<tr class="row-even"><td>RM_DUP</td>
<td>Remove duplicated reads&#8217; pairs. <em>Default: 1</em></td>
</tr>
<tr class="row-odd"><td>BIN_SIZE</td>
<td>Resolution of contact maps to generate (space separated). <em>Default: 20000 40000 150000 500000 1000000</em></td>
</tr>
<tr class="row-even"><td>BIN_STEP</td>
<td>Binning step size in ‘n’ coverage _i.e._ window step. <em>Default: 1</em></td>
</tr>
<tr class="row-odd"><td>MATRIX_FORMAT</td>
<td>Output matrix format. Must be complete, asis, upper or lower. <em>Default: upper</em></td>
</tr>
<tr class="row-even"><td>MAX_ITER</td>
<td>Maximum number of iteration for ICE normalization. <em>Default: 100</em></td>
</tr>
<tr class="row-odd"><td>SPARSE_FILTERING</td>
<td>Define which pourcentage of bins with high sparsity should be force to zero. <em>Default: 0.02</em></td>
</tr>
<tr class="row-even"><td>EPS</td>
<td>The relative increment in the results before declaring convergence. <em>Default: 0.1</em></td>
</tr>
</tbody>
</table>
</div>
<hr class="docutils" />
<div class="section" id="run-hic-pro-in-sequential-mode">
<h2>Run HiC-Pro in sequential mode<a class="headerlink" href="#run-hic-pro-in-sequential-mode" title="Permalink to this headline">¶</a></h2>
<p>HiC-Pro can be run in a step-by-step mode.
Available steps are described in the help command.</p>
<div class="highlight-guess"><div class="highlight"><pre><span class="n">HiC</span><span class="o">-</span><span class="n">Pro</span> <span class="o">--</span><span class="n">help</span>
<span class="n">usage</span> <span class="o">:</span> <span class="n">HiC</span><span class="o">-</span><span class="n">Pro</span> <span class="o">-</span><span class="n">i</span> <span class="n">INPUT</span> <span class="o">-</span><span class="n">o</span> <span class="n">OUTPUT</span> <span class="o">-</span><span class="n">c</span> <span class="n">CONFIG</span> <span class="p">[</span><span class="o">-</span><span class="n">s</span> <span class="n">ANALYSIS_STEP</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">p</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">h</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">v</span><span class="p">]</span>
<span class="n">Use</span> <span class="n">option</span> <span class="o">-</span><span class="n">h</span><span class="o">|--</span><span class="n">help</span> <span class="k">for</span> <span class="n">more</span> <span class="n">information</span>

<span class="n">HiC</span><span class="o">-</span><span class="n">Pro</span> <span class="mf">2.5.2</span>
<span class="o">---------------</span>
<span class="n">OPTIONS</span>

 <span class="o">-</span><span class="n">i</span><span class="o">|--</span><span class="n">input</span> <span class="n">INPUT</span> <span class="o">:</span> <span class="n">input</span> <span class="n">data</span> <span class="n">folder</span><span class="p">;</span> <span class="n">Must</span> <span class="n">contains</span> <span class="n">a</span> <span class="n">folder</span> <span class="n">per</span> <span class="n">sample</span> <span class="n">with</span> <span class="n">fastq</span> <span class="p">(</span><span class="n">or</span> <span class="n">bam</span><span class="p">)</span> <span class="n">files</span>
 <span class="o">-</span><span class="n">o</span><span class="o">|--</span><span class="n">output</span> <span class="n">OUTPUT</span> <span class="o">:</span> <span class="n">output</span> <span class="n">folder</span>
 <span class="o">-</span><span class="n">c</span><span class="o">|--</span><span class="n">conf</span> <span class="n">CONFIG</span> <span class="o">:</span> <span class="n">configuration</span> <span class="n">file</span> <span class="k">for</span> <span class="n">Hi</span><span class="o">-</span><span class="n">C</span> <span class="n">processing</span>
 <span class="p">[</span><span class="o">-</span><span class="n">p</span><span class="o">|--</span><span class="n">parallel</span><span class="p">]</span> <span class="o">:</span> <span class="k">if</span> <span class="n">specified</span> <span class="n">run</span> <span class="n">HiC</span><span class="o">-</span><span class="n">Pro</span> <span class="n">in</span> <span class="n">PBS</span><span class="o">/</span><span class="n">Torque</span> <span class="n">mode</span>
 <span class="p">[</span><span class="o">-</span><span class="n">s</span><span class="o">|--</span><span class="n">step</span> <span class="n">ANALYSIS_STEP</span><span class="p">]</span> <span class="o">:</span> <span class="n">run</span> <span class="n">only</span> <span class="n">a</span> <span class="n">subset</span> <span class="n">of</span> <span class="n">the</span> <span class="n">HiC</span><span class="o">-</span><span class="n">Pro</span> <span class="n">workflow</span><span class="p">;</span> <span class="k">if</span> <span class="n">not</span> <span class="n">specified</span> <span class="n">the</span> <span class="n">complete</span> <span class="n">workflow</span> <span class="n">is</span> <span class="n">run</span>
    <span class="nl">mapping:</span> <span class="n">perform</span> <span class="n">reads</span> <span class="n">alignment</span>
    <span class="nl">proc_hic:</span> <span class="n">perform</span> <span class="n">Hi</span><span class="o">-</span><span class="n">C</span> <span class="n">filtering</span>
    <span class="nl">quality_checks:</span> <span class="n">run</span> <span class="n">Hi</span><span class="o">-</span><span class="n">C</span> <span class="n">quality</span> <span class="n">control</span> <span class="n">plots</span>
    <span class="nl">build_contact_maps:</span> <span class="n">build</span> <span class="n">raw</span> <span class="n">inter</span><span class="o">/</span><span class="n">intrachromosomal</span> <span class="n">contact</span> <span class="n">maps</span>
    <span class="n">ice_norm</span> <span class="o">:</span> <span class="n">run</span> <span class="n">ICE</span> <span class="n">normalization</span> <span class="n">on</span> <span class="n">contact</span> <span class="n">maps</span>
 <span class="p">[</span><span class="o">-</span><span class="n">h</span><span class="o">|--</span><span class="n">help</span><span class="p">]</span><span class="o">:</span> <span class="n">help</span>
 <span class="p">[</span><span class="o">-</span><span class="n">v</span><span class="o">|--</span><span class="n">version</span><span class="p">]</span><span class="o">:</span> <span class="n">version</span>
</pre></div>
</div>
<p>As an exemple, if you want to only want to align the sequencing reads, use :</p>
<div class="highlight-guess"><div class="highlight"><pre><span class="n">MY_INSTALL_PATH</span><span class="o">/</span><span class="n">bin</span><span class="o">/</span><span class="n">HiC</span><span class="o">-</span><span class="n">Pro</span> <span class="o">-</span><span class="n">i</span> <span class="n">FULL_PATH_TO_RAW_DATA</span> <span class="o">-</span><span class="n">o</span> <span class="n">FULL_PATH_TO_OUTPUTS</span> <span class="o">-</span><span class="n">c</span> <span class="n">MY_LOCAL_CONFIG_FILE</span> <span class="o">-</span><span class="n">s</span> <span class="n">mapping</span> <span class="o">-</span><span class="n">s</span> <span class="n">quality_checks</span>
</pre></div>
</div>
<p>Note that in sequential mode, the INPUT argument depends on the analysis steps.</p>
<table border="1" class="docutils">
<colgroup>
<col width="53%" />
<col width="47%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head" colspan="2">INPUT DATA TYPE IN STEPWISE MODE</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>-s mapping</td>
<td>fastq(.gz) files</td>
</tr>
<tr class="row-odd"><td>-s proc_hic</td>
<td>.bam files</td>
</tr>
<tr class="row-even"><td>-s quality_checks</td>
<td>.bam files</td>
</tr>
<tr class="row-odd"><td>-s build_contact_maps</td>
<td>.validPairs files</td>
</tr>
<tr class="row-even"><td>-s ice_norm</td>
<td>.matrix files</td>
</tr>
</tbody>
</table>
<p>See te <a class="reference external" href="USER_CASES.rst">user&#8217;s cases</a> for more examples.</p>
</div>
<div class="section" id="how-does-hic-pro-work">
<h2>How does HiC-Pro work ?<a class="headerlink" href="#how-does-hic-pro-work" title="Permalink to this headline">¶</a></h2>
<div class="figure">
<a class="reference internal image-reference" href="_images/hicpro_wkflow.png"><img alt="_images/hicpro_wkflow.png" src="_images/hicpro_wkflow.png" style="width: 471.2px; height: 640.0px;" /></a>
</div>
<ol class="arabic simple">
<li><strong>Reads Mapping</strong></li>
</ol>
<p>Each mate is independantly aligned on the reference genome. The mapping is performed in two steps. First, the reads are aligned using an end-to-end aligner. Second, reads spanning the ligation junction are trimmmed from their 3&#8217; end, and aligned on the genome. Aligned reads for both fragment mates are then paired in a single paired-end BAM file. Singletons and multi-hits can be discarded according the confirguration parameters.</p>
<ol class="arabic simple" start="2">
<li><strong>Fragment assignment and filtering</strong></li>
</ol>
<p>Each aligned reads can be assigned to one restriction fragment according to the reference genome and the restriction enzyme.
The next step is to separate the invalid ligation products from the valid pairs. Dangling end and self circles pairs are therefore excluded.
Only valid pairs involving two different restriction fragments are used to build the contact maps. Duplicated valid pairs associated to PCR artefacts are discarded.
The fragment assignment can be visualized through a BAM files of aliged pairs where each pair is flagged according to its classification.</p>
<ol class="arabic simple" start="3">
<li><strong>Quality Controls</strong></li>
</ol>
<p>HiC-Pro performs a couple of quality controls for most of the analysis steps. The alignment statistics are the first quality controls. Aligned reads in the first (end-to-end) step, and alignment after trimming are reported. Note that in pratice, we ususally observed around 10-20% of trimmed reads. An abnormal level of trimmed reads can reflect a ligation issue.
Once the reads are aligned on the genome, HiC-pro checks the number of singleton, multiple hits or duplicates. The fraction of valid pairs are presented for each type of ligation products. Invalid pairs such as dangling and or self-circle are also represented. A high level of dangling ends, or an imbalance in valid pairs ligation type can be due to a ligation, fill-in or digestion issue.
Finally HiC-Pro also calculated the distribution of fragment size on a subset of valid pairs. Additional statistics will report the fraction of intra/inter-chromosomal contacts, as well as the proportion of short range (&lt;20kb) versus long range (&gt;20kb) contacts.</p>
<ol class="arabic simple" start="4">
<li><strong>Map builder</strong></li>
</ol>
<p>Intra et inter-chromosomal contact maps are build for all specified resolutions. The genome is splitted into bins of equal size. Each valid interaction is associated with the genomic bins to generate the raw maps.</p>
<ol class="arabic simple" start="5">
<li><strong>ICE normalization</strong></li>
</ol>
<p>Hi-C data can contain several sources of biases which has to be corrected. HiC-Pro proposes a fast implementation of the original ICE normalization algorithm (Imakaev et al. 2012), making the assumption of equal visibility of each fragment. The ICE normalization can be used as a standalone python package and is available <a class="reference external" href="https://github.com/hiclib/">https://github.com/hiclib/</a></p>
</div>
<div class="section" id="browsing-the-results">
<h2>Browsing the results<a class="headerlink" href="#browsing-the-results" title="Permalink to this headline">¶</a></h2>
<ul class="simple">
<li><em>bowtie_results</em></li>
<li><em>hic_results</em></li>
</ul>
<p>A contact map is defined by :</p>
<ul class="simple">
<li>A list of genomic intervals related to the specified resolution (BED format).</li>
<li>A matrix, stored as standard triplet sparse format (i.e. list format). Based on the observation that a contact map is symmetric and usually sparse, only non-zero values are stored for half of the matrix. The user can specified if the <em>&#8216;upper&#8217;</em>, <em>&#8216;lower&#8217;</em> or <em>&#8216;complete&#8217;</em> matrix has to be stored. The <em>&#8216;asis&#8217;</em> option allows to store the contacts as they are observed from the valid pairs files.</li>
</ul>
<div class="highlight-python"><div class="highlight"><pre>A   B   10
A   C   23
B   C   24
(...)
</pre></div>
</div>
<p>This format is memory efficient, and is compatible with other analysis softwares such as the <a class="reference external" href="http://bioconductor.org/packages/release/bioc/html/HiTC.html">HiTC Bioconductor package</a>.</p>
</div>
</div>


    </div>
<footer class="footer">
    <p class="pull-right">
      <a href="#">Back to top</a>
      
    </p>
    <p>
        &copy; Copyright 2015, Nicolas Servant.<br/>
      Created using <a href="http://sphinx.pocoo.org/">Sphinx</a> 1.2.2.<br/>
    </p>
</footer>
  </body>
</html>