<!DOCTYPE html>


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
  
    <title>HiC-Pro: An optimized and flexible pipeline for Hi-C processing. &mdash; HiC-Pro 2.3.1 documentation</title>
  <!-- htmltitle is before nature.css - we use this hack to load bootstrap first -->
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <link rel="stylesheet"
        href="_static/css/bootstrap.min.css"
	media="screen" />
    <script src="_static/js/jquery.js"
          type="text/javascript"></script>
  <script src="_static/js/bootstrap.min.js"
          type="text/javascript"></script>


    
    <link rel="stylesheet" href="_static/basic.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    './',
        VERSION:     '2.3.1',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <link rel="top" title="HiC-Pro 2.3.1 documentation" href="index.html" />
    <link rel="next" title="COMPATIBILITY WITH OTHER SOFTWARE" href="COMPATIBILITY.html" />
    <link rel="prev" title="HiC-Pro Utilities" href="UTILS.html" />
<meta charset='utf-8'>
<meta http-equiv='X-UA-Compatible' content='IE=edge,chrome=1'>
<meta name='viewport'
      content='width=device-width, initial-scale=1.0, maximum-scale=1'>
<meta name="apple-mobile-web-app-capable" content="yes">
<link rel="stylesheet"
        href="_static/css/jhepc.css"
         media="screen" />
  <link href='http://fonts.googleapis.com/css?family=PT+Serif'
        rel='stylesheet' type='text/css'>
  <link href='http://fonts.googleapis.com/css?family=PT+Sans'
        rel='stylesheet' type='text/css'>
  <link href='http://fonts.googleapis.com/css?family=PT+Sans+Narrow'
        rel='stylesheet' type='text/css'>



  </head>
  <body><div class="navbar">
    <a class="navbar-brand" href="index.html">
      
    </a>
    <ul class="nav navbar-nav">

    </ul>
  </div>
    <div class="container">
  
  <div class="section" id="user-cases">
<span id="usercases"></span><h1>User cases<a class="headerlink" href="#user-cases" title="Permalink to this headline">¶</a></h1>
<p>Let&#8217;s defined a few variables</p>
<ul class="simple">
<li>TEST_DATA is my input folder with one folder per sample and all my fastq files</li>
<li>RES_PREFIX is the prefix of my output directory</li>
<li>config_test.txt is my configuration file</li>
</ul>
<div class="section" id="one-mode-for-one-environment">
<h2>One mode for one environment<a class="headerlink" href="#one-mode-for-one-environment" title="Permalink to this headline">¶</a></h2>
<ol class="arabic simple">
<li><strong>Running HiC-Pro&#8217;s complete pipeline on my laptop</strong></li>
</ol>
<div class="highlight-guess"><div class="highlight"><pre>HiC-Pro -i <span class="cp">${</span><span class="n">TEST_DATA</span><span class="cp">}</span> -o <span class="cp">${</span><span class="n">RES_PREFIX</span><span class="cp">}</span>_1 -c config_test.txt
</pre></div>
</div>
<ol class="arabic simple" start="2">
<li><strong>Running HiC-Pro on a cluster with PBS/TORQUE</strong></li>
</ol>
<div class="highlight-guess"><div class="highlight"><pre>HiC-Pro -i <span class="cp">${</span><span class="n">TEST_DATA</span><span class="cp">}</span> -o <span class="cp">${</span><span class="n">RES_PREFIX</span><span class="cp">}</span>_2 -c config_test.txt -p
cd <span class="cp">${</span><span class="n">TEST_DATA</span><span class="cp">}</span>
qsub HiCPro_step1_test.sh
## wait ...
qsub HiCPro_step2_test.sh
</pre></div>
</div>
<ol class="arabic simple" start="3">
<li><strong>Running HiC-Pro using stepwise function</strong></li>
</ol>
<p>I first want to align to raw reads and make the quality controls on the mapping results using the parallel mode</p>
<div class="highlight-guess"><div class="highlight"><pre>HiC-Pro -i <span class="cp">${</span><span class="n">TEST_DATA</span><span class="cp">}</span> -o <span class="cp">${</span><span class="n">RES_PREFIX</span><span class="cp">}</span>_3 -c config_test.txt -s mapping -s quality_checks -p
cd <span class="cp">${</span><span class="n">RES_PREFIX</span><span class="cp">}</span>_3
qsub HiCPro_step1_mappingtest.sh
</pre></div>
</div>
<p>Once the reads are mapped, let&#8217;s run the Hi-C processing step from the aligned data on my own laptop</p>
<div class="highlight-guess"><div class="highlight"><pre>HiC-Pro -i <span class="cp">${</span><span class="n">RES_PREFIX</span><span class="cp">}</span>_3/bowtie_results/bwt2 -o <span class="cp">${</span><span class="n">RES_PREFIX</span><span class="cp">}</span>_3.1 -c config_test.txt -s proc_hic -s quality_checks
</pre></div>
</div>
<p>Everything looks good, I can then create the contact maps and normalize them</p>
<div class="highlight-guess"><div class="highlight"><pre>HiC-Pro -i <span class="cp">${</span><span class="n">RES_PREFIX</span><span class="cp">}</span>_3.1/hic_results/data -o <span class="cp">${</span><span class="n">RES_PREFIX</span><span class="cp">}</span>_3.2 -c config_test.txt -s build_contact_maps -s ice_norm
</pre></div>
</div>
</div>
<div class="section" id="supported-protocols">
<h2>Supported protocols<a class="headerlink" href="#supported-protocols" title="Permalink to this headline">¶</a></h2>
<ol class="arabic simple" start="4">
<li><strong>Running HiC-Pro allele specific analysis</strong></li>
</ol>
<p>First set the ALLELE_SPECIFIC_SNP variable in the configuration file and check whether the N-masked genome is available. Then simply run HiC-Pro as usually.
See the <a class="reference internal" href="AS.html#as"><em>allele specific</em></a> section for details.</p>
<div class="highlight-guess"><div class="highlight"><pre>HiC-Pro -i <span class="cp">${</span><span class="n">TEST_DATA</span><span class="cp">}</span> -o <span class="cp">${</span><span class="n">RES_PREFIX</span><span class="cp">}</span>_5 -c config_test_as.txt
</pre></div>
</div>
<ol class="arabic simple" start="5">
<li><strong>Running HiC-Pro on DNase Hi-C data</strong></li>
</ol>
<p>HiC-Pro is able to process Hi-C data generated without restriction enzyme, such as DNase Hi-C.
To analyse such data, simply unset the variable LIGATION_SITE and GENOME_FRAGMENT from the configuration file.
As the ligation site is unknown, the mapping will be done in only one step (step 2 is not run). Then, the uniquely mapped reads are directly used to build the contact maps.
In the context of DNase Hi-C, one way to filter out artefacts such as self ligation is to discard intra-chromosomal pairs below a given distance threshold.
A filtering of the short contact distance is therefore proposed using the parameter MIN_CIS_DIST.</p>
<ol class="arabic simple" start="6">
<li>** Analysis of capture-C data**</li>
</ol>
<p>In practice, the processing of capture-C data has many common steps with standard Hi-C analyis. HiC-Pro can therefore be used for capture-C data until the detection of valid 3C products. You should therefore used HiC-Pro is stepwise mode with option &#8220;-s mapping -s proc_hic -s quality_checks -s merge_persample&#8221;.
One the list of valid pairs available, we proposed an additional utility &#8216;make_viewpoints.py&#8217; which is able to build bedgraph file for a given list of viewpoints.</p>
<ol class="arabic simple" start="7">
<li>** Analysis of capture Hi-C data**</li>
</ol>
<p>HiC-Pro can also be used to process capture Hi-C data. In this case, you have to provide a BED file with the target regions and to set the variable CAPTURE_TARGET in the config file. After mapping and filtering, the valid pairs will be restricted to the intra-chromosomal interactions that occur within the designed regions. All other interaction will be removed. The interaction maps will therefore be build on that region.</p>
</div>
</div>


    </div>
<footer class="footer">
    <p class="pull-right">
      <a href="#">Back to top</a>
      
    </p>
    <p>
        &copy; Copyright 2015, Nicolas Servant.<br/>
      Created using <a href="http://sphinx.pocoo.org/">Sphinx</a> 1.2.2.<br/>
    </p>
</footer>
  </body>
</html>