<!DOCTYPE html>


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
  
    <title>HiC-Pro: An optimized and flexible pipeline for Hi-C processing. &mdash; HiC-Pro 2.3.1 documentation</title>
  <!-- htmltitle is before nature.css - we use this hack to load bootstrap first -->
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <link rel="stylesheet"
        href="_static/css/bootstrap.min.css"
	media="screen" />
    <script src="_static/js/jquery.js"
          type="text/javascript"></script>
  <script src="_static/js/bootstrap.min.js"
          type="text/javascript"></script>


    
    <link rel="stylesheet" href="_static/basic.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    './',
        VERSION:     '2.3.1',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <link rel="top" title="HiC-Pro 2.3.1 documentation" href="index.html" />
    <link rel="next" title="User cases" href="USER_CASES.html" />
    <link rel="prev" title="HiC-Pro Results" href="RESULTS.html" />
<meta charset='utf-8'>
<meta http-equiv='X-UA-Compatible' content='IE=edge,chrome=1'>
<meta name='viewport'
      content='width=device-width, initial-scale=1.0, maximum-scale=1'>
<meta name="apple-mobile-web-app-capable" content="yes">
<link rel="stylesheet"
        href="_static/css/jhepc.css"
         media="screen" />
  <link href='http://fonts.googleapis.com/css?family=PT+Serif'
        rel='stylesheet' type='text/css'>
  <link href='http://fonts.googleapis.com/css?family=PT+Sans'
        rel='stylesheet' type='text/css'>
  <link href='http://fonts.googleapis.com/css?family=PT+Sans+Narrow'
        rel='stylesheet' type='text/css'>



  </head>
  <body role="document"><div class="navbar">
    <a class="navbar-brand" href="index.html">
      
    </a>
    <ul class="nav navbar-nav">

    </ul>
  </div>
    <div class="container">
  
  <div class="section" id="hic-pro-utilities">
<h1>HiC-Pro Utilities<a class="headerlink" href="#hic-pro-utilities" title="Permalink to this headline">¶</a></h1>
<div class="line-block" id="utils">
<div class="line">HiC-Pro provides a couple of utilities which are not part of the main pipeline, but was designed to help the user to perform a couple of tasks.</div>
<div class="line">All utilities are installed in /bin/utils/.</div>
</div>
<div class="section" id="split-reads-py">
<h2>1- split_reads.py<a class="headerlink" href="#split-reads-py" title="Permalink to this headline">¶</a></h2>
<p>or <strong>How can I split my reads in chunck ?</strong></p>
<div class="line-block">
<div class="line">Split reads is highly recommanded in HiC-pro parallel mode.</div>
<div class="line">To split the reads in chunck, you can simply used the split_reads.py utility.</div>
</div>
<div class="highlight-bash"><div class="highlight"><pre><span class="c">## split the fastq file in blocks of 10M reads</span>
HICPRO_PATH/bin/utils/split_reads.py --results_folder OUTPUT --nreads READS_NB INPUT_FASTQ
</pre></div>
</div>
</div>
<div class="section" id="extract-snps-py">
<h2>2- extract_snps.py<a class="headerlink" href="#extract-snps-py" title="Permalink to this headline">¶</a></h2>
<p>or <strong>How can I extract SNPs information from phasing data ?</strong></p>
<div class="line-block">
<div class="line">This script was designed to extract SNPs information from a VCF file.</div>
<div class="line">As a example, the Sanger database provides a list of SNPs which can be used to differentiate several Mouse strains (see <a class="reference external" href="ftp://ftp-mouse.sanger.ac.uk/current_snps/">ftp://ftp-mouse.sanger.ac.uk/current_snps/</a>).</div>
<div class="line">The single VCF contains all SNPs information for all Mouse strains. The default reference allele is the C57 black strain.</div>
<div class="line">This script is able to extract informative and high quality SNPs (-f option) which can be then used in allele specific analysis to distinguish parental origin.</div>
</div>
<div class="highlight-bash"><div class="highlight"><pre><span class="c">## Extract SNPs information for CASTEiJ/129S1 cross</span>
HICPRO_PATH/bin/utils/split_reads.py -i mgp.v2.snps.annot.reformat.vcf -r CASTEij -a 129S1 &gt; snps_CASTEiJ_129S1.vcf

<span class="c">## Extract SNPs information for C57_b6/FVB_NJ</span>
HICPRO_PATH/bin/utils/split_reads.py -i mgp.v2.snps.annot.reformat.vcf -a FVB_NJ &gt; snps_CASTEiJ_129S1.vcf
</pre></div>
</div>
</div>
<div class="section" id="digest-genome-py">
<h2>2- digest_genome.py<a class="headerlink" href="#digest-genome-py" title="Permalink to this headline">¶</a></h2>
<p>or <strong>How can I generate the list of restriction fragments after genome digestion ?</strong></p>
<div class="line-block">
<div class="line">Digest the reference genome by the provided restriction enzymes(s) and generate a BED file with the list of restriction fragments after digestion.</div>
<div class="line">This file can then be used by HiC-Pro (GENOME_FRAGMENT) for the data processing.</div>
<div class="line">Note that the cutting site of the restriction enzyme has to be specified using the &#8216;^&#8217; character.</div>
<div class="line">The restriction enzymes HindIII, DpnII and BglII are encoded within the script and are therefore recognized if specified to the program.</div>
<div class="line">Finally, multiple restriction enzymes can also be provided.</div>
</div>
<div class="highlight-bash"><div class="highlight"><pre><span class="c">## Digest the mm9 genome by HindIII</span>
HICPRO_PATH/bin/utils/digest_genome.py -r A^AGCTT -o mm9_hindiii.bed mm9.fasta

<span class="c">## The same ...</span>
HICPRO_PATH/bin/utils/digest_genome.py -r hindiii -o mm9_hindiii.bed mm9.fasta

<span class="c">## Double digestion, HindIII + DpnII</span>
HICPRO_PATH/bin/utils/digest_genome.py -r hindiii dpnii -o mm9_hindiii_dpnii.bed mm9.fasta
</pre></div>
</div>
</div>
</div>


    </div>
<footer class="footer">
    <p class="pull-right">
      <a href="#">Back to top</a>
      
    </p>
    <p>
        &copy; Copyright 2015, Nicolas Servant.<br/>
      Created using <a href="http://sphinx.pocoo.org/">Sphinx</a> 1.3.1.<br/>
    </p>
</footer>
  </body>
</html>