<!DOCTYPE html>


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
  
    <title>HiC-Pro: An optimized and flexible pipeline for Hi-C processing. &mdash; HiC-Pro 2.3.1 documentation</title>
  <!-- htmltitle is before nature.css - we use this hack to load bootstrap first -->
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <link rel="stylesheet"
        href="_static/css/bootstrap.min.css"
	media="screen" />
    <script src="_static/js/jquery.js"
          type="text/javascript"></script>
  <script src="_static/js/bootstrap.min.js"
          type="text/javascript"></script>


    
    <link rel="stylesheet" href="_static/basic.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    './',
        VERSION:     '2.3.1',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <link rel="top" title="HiC-Pro 2.3.1 documentation" href="index.html" />
    <link rel="next" title="HiC-Pro Results" href="RESULTS.html" />
    <link rel="prev" title="HiC-Pro Manual" href="MANUAL.html" />
<meta charset='utf-8'>
<meta http-equiv='X-UA-Compatible' content='IE=edge,chrome=1'>
<meta name='viewport'
      content='width=device-width, initial-scale=1.0, maximum-scale=1'>
<meta name="apple-mobile-web-app-capable" content="yes">
<link rel="stylesheet"
        href="_static/css/jhepc.css"
         media="screen" />
  <link href='http://fonts.googleapis.com/css?family=PT+Serif'
        rel='stylesheet' type='text/css'>
  <link href='http://fonts.googleapis.com/css?family=PT+Sans'
        rel='stylesheet' type='text/css'>
  <link href='http://fonts.googleapis.com/css?family=PT+Sans+Narrow'
        rel='stylesheet' type='text/css'>



  </head>
  <body><div class="navbar">
    <a class="navbar-brand" href="index.html">
      
    </a>
    <ul class="nav navbar-nav">

    </ul>
  </div>
    <div class="container">
  
  <div class="section" id="allele-specific-analysis">
<h1>Allele specific analysis<a class="headerlink" href="#allele-specific-analysis" title="Permalink to this headline">Â¶</a></h1>
<p>If the ALLELE_SPECIFIC_SNP option is defined in the configuration, HiC-Pro will run the allele specific mode.
The ALLELE_SPECIFIC_SNP option must contain the path to a VCF file with phasing SNPs information.
HiC-Pro provides the <a class="reference external" href="UTILS.rst">*extract_snps.py* utilities</a> to generate such file.
This utility was first design to extract relevant information from the Mouse Sanger database SNPs file. The goal is to generate a VCF file of the F1 individual according to its parental genotype. For instance using <em>extract_snps.py</em> with <em>-r CASTEiJ</em> and <em>-a 129S1</em> will generate a VCF file with all F1 heterogyzote SNPs which can be used to distinguish <em>CASTEiJ</em> and <em>129S1</em> alleles.</p>
<p><strong>Allele specific mapping</strong></p>
<p>In allele-specific mode, the sequencing reads are first aligned on masked reference genome for which all polymorphic sites were first N-masked.
In the current version, <strong>this is the user responsability to provide to generate this genome and to provide the bowtie2 indexes to HiC-Pro.</strong>
Example of how to generate such reference genome is discribed in the <a class="reference external" href="FAQ.rst">FAQ</a> section.
This masking strategy avoid systematic bias toward the reference allele, compared to standard mapping where reads with the reference allele are more likely to be mapped than the reads with non-reference alleles.</p>
<p><strong>Assignment to parental genome</strong></p>
<p>Once aligned HiC-Pro browses all reads spanning a polymorphic site, locates the nucleotide at the appropriate position, and assigns the read either to the maternal or paternal allele. Reads with conflicting allele assignment or unexpected allele at polymorphic sites are discarded.
This step generates a bam file with all reads flagged according to its parental assignment (AS flag).
<strong>AS:i:0</strong> - unassigned (UA)
<strong>AS:i:1</strong> - assigned to reference genome (G1)
<strong>AS:i:2</strong> - assigned to alternative genome (G2)
<strong>AS:i:3</strong> - conflicting (C)</p>
<p><strong>Allele specific maps</strong></p>
<p>Finally, read pairs with at least one allele specific mate are used to construct allele specific contact maps.
In allele-specific mode, the <em>.validPairs</em> file will contains an additional columns with the allele status of each reads pair.
Therefore, valid pairs assigned to &#8216;G1-G1&#8217;, &#8216;G1-UA&#8217; or &#8216;UA-G1&#8217; are used to build the G1 genome-wide contact maps. In the same way, &#8216;G2-G2&#8217;, &#8216;G2-UA&#8217;, &#8216;UA-G2&#8217; pairs will be used to generate the G2 genome-wide contact maps.
All maps are then normalized using the ICE method.</p>
</div>


    </div>
<footer class="footer">
    <p class="pull-right">
      <a href="#">Back to top</a>
      
    </p>
    <p>
        &copy; Copyright 2015, Nicolas Servant.<br/>
      Created using <a href="http://sphinx.pocoo.org/">Sphinx</a> 1.2.2.<br/>
    </p>
</footer>
  </body>
</html>