<!DOCTYPE html>


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
  
    <title>HiC-Pro: An optimized and flexible pipeline for Hi-C processing. &mdash; HiC-Pro 2.3.1 documentation</title>
  <!-- htmltitle is before nature.css - we use this hack to load bootstrap first -->
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <link rel="stylesheet"
        href="_static/css/bootstrap.min.css"
	media="screen" />
    <script src="_static/js/jquery.js"
          type="text/javascript"></script>
  <script src="_static/js/bootstrap.min.js"
          type="text/javascript"></script>


    
    <link rel="stylesheet" href="_static/basic.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    './',
        VERSION:     '2.3.1',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <link rel="top" title="HiC-Pro 2.3.1 documentation" href="index.html" />
    <link rel="next" title="HiC-Pro Utilities" href="UTILS.html" />
    <link rel="prev" title="Allele specific analysis" href="AS.html" />
<meta charset='utf-8'>
<meta http-equiv='X-UA-Compatible' content='IE=edge,chrome=1'>
<meta name='viewport'
      content='width=device-width, initial-scale=1.0, maximum-scale=1'>
<meta name="apple-mobile-web-app-capable" content="yes">
<link rel="stylesheet"
        href="_static/css/jhepc.css"
         media="screen" />
  <link href='http://fonts.googleapis.com/css?family=PT+Serif'
        rel='stylesheet' type='text/css'>
  <link href='http://fonts.googleapis.com/css?family=PT+Sans'
        rel='stylesheet' type='text/css'>
  <link href='http://fonts.googleapis.com/css?family=PT+Sans+Narrow'
        rel='stylesheet' type='text/css'>



  </head>
  <body><div class="navbar">
    <a class="navbar-brand" href="index.html">
      
    </a>
    <ul class="nav navbar-nav">

    </ul>
  </div>
    <div class="container">
  
  <div class="section" id="hic-pro-results">
<h1>HiC-Pro Results<a class="headerlink" href="#hic-pro-results" title="Permalink to this headline">¶</a></h1>
<div class="section" id="mapping-results">
<h2>Mapping results<a class="headerlink" href="#mapping-results" title="Permalink to this headline">¶</a></h2>
<p>The <em>bowtie_results</em> folder contains the results of the reads mapping. The results of first mapping step are available in the <em>bwt2_glob</em> folder, and the seconnd step in the <em>bwt2_loc</em> folder. Final BAM files, reads pairing, and mapping statistics are available on the <em>bwt2</em> folder.</p>
</div>
<div class="section" id="list-of-valid-interaction-products">
<h2>List of valid interaction products<a class="headerlink" href="#list-of-valid-interaction-products" title="Permalink to this headline">¶</a></h2>
<p>The <em>hic_results</em> folder contains all Hi-C processed data, and is further divided in several sub-folders.
The data folder is used to store the valid interaction products (.validPairs), as well as other statisics files.</p>
</div>
<div class="section" id="intra-and-inter-chromosomal-contact-maps">
<h2>Intra and inter-chromosomal contact maps<a class="headerlink" href="#intra-and-inter-chromosomal-contact-maps" title="Permalink to this headline">¶</a></h2>
<p>The contact maps are then available in the <em>matrix</em> folder.</p>
<p>A contact map is defined by :</p>
<ul class="simple">
<li>A list of genomic intervals related to the specified resolution (BED format).</li>
<li>A matrix, stored as standard triplet sparse format (i.e. list format). Based on the observation that a contact map is symmetric and usually sparse, only non-zero values are stored for half of the matrix. The user can specified if the <em>&#8216;upper&#8217;</em>, <em>&#8216;lower&#8217;</em> or <em>&#8216;complete&#8217;</em> matrix has to be stored. The <em>&#8216;asis&#8217;</em> option allows to store the contacts as they are observed from the valid pairs files.</li>
</ul>
<div class="highlight-python"><div class="highlight"><pre>A   B   10
A   C   23
B   C   24
(...)
</pre></div>
</div>
<p>This format is memory efficient, and is compatible with other analysis softwares such as the <a class="reference external" href="http://bioconductor.org/packages/release/bioc/html/HiTC.html">HiTC Bioconductor package</a>. The <em>matrix</em> folder is organized with <em>raw</em> and <em>iced</em> contact maps for all resolutions.</p>
</div>
</div>


    </div>
<footer class="footer">
    <p class="pull-right">
      <a href="#">Back to top</a>
      
    </p>
    <p>
        &copy; Copyright 2015, Nicolas Servant.<br/>
      Created using <a href="http://sphinx.pocoo.org/">Sphinx</a> 1.2.2.<br/>
    </p>
</footer>
  </body>
</html>